File: schemas/delta.py
"""
VAEL – Delta / Monitoring Schemas

Represents changes between two VAEL analyses of the same software version,
run at different points in time.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field

from schemas.stage2 import VEPTier, ExploitMaturity


class ChangeType(str, Enum):
    NEW_CVE         = "NEW_CVE"          # CVE not present in baseline
    REMOVED_CVE     = "REMOVED_CVE"      # CVE in baseline but not in current
    TIER_UPGRADE    = "TIER_UPGRADE"     # VEP tier moved to higher severity
    TIER_DOWNGRADE  = "TIER_DOWNGRADE"   # VEP tier moved to lower severity
    KEV_ADDED       = "KEV_ADDED"        # CVE newly added to CISA KEV
    KEV_REMOVED     = "KEV_REMOVED"      # rare
    EPSS_SPIKE      = "EPSS_SPIKE"       # EPSS score increased by ≥ 0.10
    EPSS_DROP       = "EPSS_DROP"        # EPSS score dropped by ≥ 0.10
    MATURITY_CHANGE = "MATURITY_CHANGE"  # exploit maturity upgraded
    NEW_POC         = "NEW_POC"          # new public PoC discovered
    NEW_WEAPON      = "NEW_WEAPON"       # weaponized PoC appeared


class CVEDelta(BaseModel):
    cve_id: str
    change_type: ChangeType
    severity: str = ""            # for sorting/display
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    detail: str = ""

    @property
    def is_high_signal(self) -> bool:
        return self.change_type in (
            ChangeType.NEW_CVE,
            ChangeType.TIER_UPGRADE,
            ChangeType.KEV_ADDED,
            ChangeType.EPSS_SPIKE,
            ChangeType.NEW_WEAPON,
            ChangeType.MATURITY_CHANGE,
        )


class DeltaReport(BaseModel):
    """Diff between two VAEL runs for the same software@version."""
    software: str
    version: str

    baseline_ts: Optional[datetime] = None
    current_ts: datetime = Field(default_factory=datetime.utcnow)

    changes: list[CVEDelta] = Field(default_factory=list)

    # Summary counts
    new_cves: int = 0
    removed_cves: int = 0
    tier_upgrades: int = 0
    kev_additions: int = 0
    epss_spikes: int = 0
    new_pocs: int = 0
    new_weapons: int = 0

    # Convenience
    high_signal_changes: list[CVEDelta] = Field(default_factory=list)
    has_critical_changes: bool = False

    def compute_summary(self) -> None:
        self.new_cves       = sum(1 for c in self.changes if c.change_type == ChangeType.NEW_CVE)
        self.removed_cves   = sum(1 for c in self.changes if c.change_type == ChangeType.REMOVED_CVE)
        self.tier_upgrades  = sum(1 for c in self.changes if c.change_type == ChangeType.TIER_UPGRADE)
        self.kev_additions  = sum(1 for c in self.changes if c.change_type == ChangeType.KEV_ADDED)
        self.epss_spikes    = sum(1 for c in self.changes if c.change_type == ChangeType.EPSS_SPIKE)
        self.new_pocs       = sum(1 for c in self.changes if c.change_type == ChangeType.NEW_POC)
        self.new_weapons    = sum(1 for c in self.changes if c.change_type == ChangeType.NEW_WEAPON)
        self.high_signal_changes = [c for c in self.changes if c.is_high_signal]
        self.has_critical_changes = bool(
            self.tier_upgrades or self.kev_additions or
            self.new_weapons or self.new_cves
        )

File: web/js/delta.js
  // ── Delta Tracker ─────────────────────────────────────────────────────
  const _DELTA_BADGE = {
    NEW_CVE:        ['new-cve',    '+ NEW CVE'],
    TIER_UPGRADE:   ['tier-up',    '▲ TIER UP'],
    TIER_DOWNGRADE: ['tier-down',  '▼ TIER DOWN'],
    KEV_ADDED:      ['kev-added',  '🔴 KEV ADDED'],
    KEV_REMOVED:    ['kev-removed','KEV REMOVED'],
    EPSS_SPIKE:     ['epss-spike', '▲ EPSS SPIKE'],
    EPSS_DROP:      ['epss-drop',  '▼ EPSS DROP'],
    MATURITY_CHANGE:['maturity',   '⚠ MATURITY'],
    NEW_POC:        ['new-poc',    '+ NEW POC'],
    NEW_WEAPON:     ['new-weapon', '⚠ WEAPONIZED'],
    REMOVED_CVE:    ['removed',    '− REMOVED'],
  };

  async function _checkDeltaBaseline() {
    const sw  = _currentSoftware;
    const ver = _currentVersion;
    const info = document.getElementById('deltaBaselineInfo');
    if (!sw || !ver) {
      info.innerHTML = `<p style="color:var(--muted);font-size:0.85rem">${t('deltaNoSoftware')}</p>`;
      return;
    }
    info.innerHTML = `<p style="color:var(--muted);font-size:0.85rem">${t('deltaChecking')}</p>`;
    try {
      const r = await fetch(`/monitor/${encodeURIComponent(sw)}/${encodeURIComponent(ver)}`);
      const d = await r.json();
      if (d.has_baseline) {
        const saved = d.saved_at ? new Date(d.saved_at).toLocaleString() : '—';
        info.innerHTML = `<div style="padding:10px 14px;background:rgba(48,209,88,0.07);border:1px solid rgba(48,209,88,0.25);border-radius:var(--radius-sm);font-size:0.84rem;">
          <span style="color:var(--green);font-weight:600">${t('deltaHasBaseline')}</span>
          <span style="color:var(--muted);margin-left:12px">${sw} ${ver} · ${saved} · ${d.cve_count ?? 0} CVEs</span>
        </div>`;
      } else {
        info.innerHTML = `<div style="padding:10px 14px;background:rgba(255,159,10,0.06);border:1px solid rgba(255,159,10,0.25);border-radius:var(--radius-sm);font-size:0.84rem;color:var(--orange)">${t('deltaNoBaseline')}</div>`;
      }
    } catch (e) {
      info.innerHTML = '';
    }
  }

  async function runDeltaAnalysis() {
    const sw  = _currentSoftware || document.getElementById('software').value.trim();
    const ver = _currentVersion  || document.getElementById('version').value.trim();
    if (!sw || !ver) {
      document.getElementById('deltaOutput').innerHTML =
        `<p style="color:var(--muted);font-size:0.85rem">${t('deltaNoSoftware')}</p>`;
      return;
    }
    const btn = document.getElementById('deltaRunBtn');
    btn.disabled = true;
    document.getElementById('deltaOutput').innerHTML =
      `<p style="color:var(--muted);font-size:0.85rem">${t('deltaRunning')}</p>`;

    try {
      const body = { software: sw, version: ver };
      const r = await fetch('/analyze/delta', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });
      if (!r.ok) {
        const err = await r.json().catch(() => ({ detail: r.statusText }));
        document.getElementById('deltaOutput').innerHTML =
          `<p style="color:var(--red);font-size:0.85rem">Error: ${err.detail || r.statusText}</p>`;
        return;
      }
      const report = await r.json();
      _renderDelta(report);
      await _checkDeltaBaseline();
    } catch (e) {
      document.getElementById('deltaOutput').innerHTML =
        `<p style="color:var(--red);font-size:0.85rem">Request failed: ${e.message}</p>`;
    } finally {
      btn.disabled = false;
    }
  }

  function _renderDelta(r) {
    const out = document.getElementById('deltaOutput');
    const changes = r.changes || [];

    if (!changes.length) {
      out.innerHTML = `<p style="color:var(--green);font-size:0.9rem;padding:16px 0">${t('deltaNoChanges')}</p>`;
      return;
    }

    const baseFmt = r.baseline_ts ? new Date(r.baseline_ts).toLocaleString() : null;
    const curFmt  = r.current_ts  ? new Date(r.current_ts).toLocaleString()  : new Date().toLocaleString();

    const timeline = baseFmt
      ? `<div class="delta-timeline">
           <strong>${baseFmt}</strong>
           <span style="font-size:1rem">→</span>
           <strong>${curFmt}</strong>
         </div>`
      : `<div class="delta-timeline"><strong>First analysis — baseline established at ${curFmt}</strong></div>`;

    const highSignal = changes.filter(c =>
      ['NEW_CVE','TIER_UPGRADE','KEV_ADDED','EPSS_SPIKE','NEW_WEAPON','MATURITY_CHANGE'].includes(c.change_type)
    );

    const statItems = [
      { num: r.new_cves,       lbl: 'New CVEs',       color: r.new_cves     ? 'var(--red)'    : '' },
      { num: r.tier_upgrades,  lbl: 'Tier Upgrades',  color: r.tier_upgrades? 'var(--orange)' : '' },
      { num: r.kev_additions,  lbl: 'KEV Added',      color: r.kev_additions? 'var(--red)'    : '' },
      { num: r.epss_spikes,    lbl: 'EPSS Spikes',    color: r.epss_spikes  ? 'var(--orange)' : '' },
      { num: r.new_pocs,       lbl: 'New PoCs',       color: r.new_pocs     ? 'var(--purple)' : '' },
      { num: r.new_weapons,    lbl: 'Weaponized',     color: r.new_weapons  ? 'var(--red)'    : '' },
      { num: r.removed_cves,   lbl: 'Removed',        color: '' },
    ].map(s =>
      `<div class="delta-stat">
         <div class="num" style="${s.color ? `color:${s.color}` : 'color:var(--text)'}">${s.num ?? 0}</div>
         <div class="lbl">${s.lbl}</div>
       </div>`
    ).join('');

    const sortOrder = ['NEW_WEAPON','KEV_ADDED','TIER_UPGRADE','MATURITY_CHANGE','EPSS_SPIKE','NEW_CVE','NEW_POC','EPSS_DROP','TIER_DOWNGRADE','KEV_REMOVED','REMOVED_CVE'];
    const sorted = [...changes].sort((a,b) => sortOrder.indexOf(a.change_type) - sortOrder.indexOf(b.change_type));

    const rows = sorted.map(c => {
      const [cls, label] = _DELTA_BADGE[c.change_type] || ['new-poc', c.change_type];
      const oldNew = (c.old_value && c.new_value)
        ? `<span style="color:var(--muted)">${c.old_value} → </span><strong>${c.new_value}</strong>`
        : (c.new_value ? `<strong>${c.new_value}</strong>` : '');
      return `<div class="delta-change-row">
        <div style="min-width:90px"><span class="delta-badge ${cls}">${label}</span></div>
        <div style="flex:0 0 130px;font-family:var(--mono);font-size:0.83rem;color:var(--accent-hi)">${c.cve_id}</div>
        <div style="flex:1;color:var(--muted)">${c.detail || ''} ${oldNew}</div>
      </div>`;
    }).join('');

    const criticalBanner = r.has_critical_changes
      ? `<div style="padding:12px 16px;background:rgba(255,69,58,0.08);border:1px solid rgba(255,69,58,0.3);border-radius:var(--radius-sm);margin-bottom:20px;color:var(--red);font-weight:600;font-size:0.88rem">
           ⚠ Critical changes detected — ${highSignal.length} high-signal event${highSignal.length !== 1 ? 's' : ''} require attention
         </div>`
      : '';

    out.innerHTML = `
      ${timeline}
      ${criticalBanner}
      <div class="delta-stat-row">${statItems}</div>
      <div class="section-label" style="margin-bottom:12px">Change Log (${changes.length})</div>
      <div>${rows}</div>`;
  }
